[SYNC][ENGINE] Engine refactor and DNS sync verify signature

JackyWYX opened the pull request:
## Description

1. Refactored engine verify signature logic, and get rid of a lot of duplicate code.
2. Remove insert helper in downloader module, use engine signature verification.
3. Added signature verification in DNS sync logic, and updated corresponding protobuf message.

All changes are backward compatible.

### A few more words about engine refactor

1. Engine is now initialized via New method, and all use of global variable of Engine is removed.
2. Engine is now initialized with a shardID, which enables engine itself to determine whether to recalculate state shard when comparing shard ID (case for verifying CX receipts).
3. Added a small cache in engine to avoid duplicate signature verification to save some CPU (`verifiedSigCache`)

## Test

1. Tested locally.
2. Tested on stressnet of DNS sync.
3. Tested on stressnet of stream sync.
4. Tested on mainnet of 1. stream sync 2. epoch 0 blocks.

rlan35 commented on api/service/legacysync/syncing.go:
```diff
@@ -844,7 +873,7 @@ func (ss *StateSync) generateNewState(bc *core.BlockChain, worker *worker.Worker
 			break
 		}
-		err = ss.UpdateBlockAndStatus(block, bc, worker, false)
+		err = ss.UpdateBlockAndStatus(block, bc, worker, true)
```
why change to true for verifyAllSig?

JackyWYX replied:
These blocks were designed to be obtained from DNS node's register/push (DownloaderRequest_REGISTER). And these blocks can possibly be the last block of one sync loop after dealing with "commonBlocks". So we need to verify the signature and flush commitSigAndBitmap to db. The amount of this signature verification is not large at all.

JackyWYX replied:
This reminds to to also update NEWBLOCK request. Thanks.

JackyWYX replied:
NEWBLOCK push request also updated with Signature. So this value can be safely set to true. This is to verify the new block signature pushed from DNS nodes. The amount of this blocks is not much compared with `commonBlocks`.

JackyWYX replied:
Updated logic to fix errors. Need two bools to check the signature values. Please check the latest code again. Thanks!

rlan35 commented on api/service/legacysync/syncing.go:
```diff
@@ -822,16 +848,19 @@ func (ss *StateSync) generateNewState(bc *core.BlockChain, worker *worker.Worker
 		}
-		err = ss.UpdateBlockAndStatus(block, bc, worker, false)
+		next := ss.getBlockFromOldBlocksByParentHash(block.Hash())
+		// Enforce sig check for the last block in a batch
```
why does this mean it's last block? Also can you explain the logic that involves getBlockFromOldBlocksByParentHash? what is oldBlocks?

JackyWYX replied:
`oldBlocks` is just some legacy function names. oldBlocks (a.k.a. `ss.commonBlocks`) are blocks downloaded from `ss.downloadBlocks`, which makes up the major portion of synced blocks. E.g. For an out of sync node, one sync loop syncs 1000 blocks, and all these 1000 blocks downloaded from remote nodes resides in `ss.commonBlocks`.

This part of logic change is basically aiming to enforce signature verification (also flush commitSig to db) for the last block of downloaded blocks. E.g, A sync loop has downloaded 900 blocks from remote nodes, and 900th block is the latest block. So this block's sig should be verified and inserted into db, which enables consensus to find the commit sig of current block block.

JackyWYX replied:
I did minimum code change but correct the logics since this is the code about to discard. Do you think I shall refactor it in a more readable way?

rlan35 replied:
got it thanks

rlan35 commented on hmy/downloader/shortrange.go:
```diff
@@ -51,21 +49,25 @@ func (d *Downloader) doShortRangeSync() (int, error) {
-		if !errors.As(err, &emptySigVerifyError) {
+		if sh.blameAllStreams(blocks, n, err) {
 			sh.removeStreams(whitelist) // Data provided by remote nodes is corrupted
+		} else {
```
what's the cases when a peer is blamed? what happen after the blaming?

rlan35 replied:
?

JackyWYX replied:
1. This is the case where the peer provides the last block in a batch gives a wrong commit signature. Since the commit signature is not include in block hash, it could be the case that other whitelist streams provides the correct hash chain, and only this node is giving wrong data.
2. In other cases, since all block data is deterministic with a fixed block hash, if signature verification or block insertion fails, we can imply that the given hash chain is invalid. Thus all streams providing that hash chain is guilty.
3. Currently, for the guilty screams, we simply close the connection to that stream. Will implement a timed blacklist later.

rlan35 commented on hmy/downloader/downloader.go:
```diff
@@ -262,3 +264,53 @@ func (d *Downloader) finishSyncing() {
+	if err := bc.Engine().VerifyHeaderSignature(bc, block.Header(), sigBytes, bitmap); err != nil {
+		return &sigVerifyErr{err}
+	}
+	if err := bc.Engine().VerifyHeader(bc, block.Header(), true); err != nil {
```
why we still verifySeal==true when we already verified the signature?

JackyWYX replied:
To verify the signature of previous block. There can be the corner case that 

```
prevBlock.CommitSigAndBitmap != block.LastCommitSig + block.LastCommitBitmap
```

So we need to verify the seal against last block. 

BTW, engine will not do the same signature verification twice because of `engineImpl.verifiedSigCache` in normal cases.

rlan35 replied:
ok.

rlan35 commented on api/service/legacysync/syncing.go:
```diff
@@ -822,16 +848,19 @@ func (ss *StateSync) generateNewState(bc *core.BlockChain, worker *worker.Worker
 			break
 		}
-		parentHash = block.Hash()
+		block = next
```
parentHash is not updated. This loop doesn't seem to work anymore.

JackyWYX replied:
ParentHash is now only used once. For the rest of the loop, `block.Hash()` is used and keeps iterating.

rlan35 replied:
I see. this logic seems more convoluted than the previous one. Why change the loop logic? The previous one seems more logical for readers.

JackyWYX replied:
Because I want to check whether the current block is the last one in `commonBlocks`. Sorry for not focusing on the code quality of these codes since I thought the code was about to be discarded. Let me try refactor a little bit.

JackyWYX replied:
Just refactored. Please let me know whether the new logic makes sense to you. Thanks for pointing this out!

JackyWYX commented:
@rlan35 Updated following changes since your last approve:

1. Added commit signature in NEWBLOCK dns sync requests. Also updated REGISTER dns request to accommodate the changes to enable backward compatibility.
2. Bug fixes regarding beacon sync and CX receipt verification for travis tests. 

Please review the code again and let me know if there are still any problems. Thanks!

rlan35 commented on api/service/legacysync/syncing.go:
```diff
@@ -844,7 +905,7 @@ func (ss *StateSync) generateNewState(bc *core.BlockChain, worker *worker.Worker
 			break
 		}
-		err = ss.UpdateBlockAndStatus(block, bc, worker, false)
+		err = ss.UpdateBlockAndStatus(block, bc, true, true)
```
why turn verifyAllSeal to true here? then it will verify signature for all blocks, isn't it slow?

JackyWYX replied:
Here is only a few block to be updated in this case compared to commonBlocks. These blocks is a part of REGISTER / PUSH flow. The logics are:

1. DNS server got a new block in consensus. Block is then pushed to `VerifiedNewBlock` (consensus/consensus_v2.go:712)
2. DNS server send NEWBLOCK request with the new block (with sig) to registered sync peers. (node/node_syncing.go:383)
3. Client node receive NEWBLOCK request, and add blocks to `peerConfig.newBlocks` (api/service/legacysync/syncing.go:253)
4. Blocks added in last step is pulled from `getMaxConsensusBlockFromParentHash` (api/service/legacysync/syncing.go:904), and being handled here.

So blocks updated here are actually small in number compared to commonBlocks. Not a very critical performance issue I think. On the other hand, this enforce signature check will also flush the block's commit sig to db, which ensures that the last block synced from dns sync is committed with a commit sig for consensus to use.

rlan35 replied:
cool

rlan35 commented on node/node_syncing.go:
```diff
@@ -567,6 +595,60 @@ func (node *Node) getEncodedBlockByHash(hash common.Hash) ([]byte, error) {
+	if blk == nil {
+		return nil, errBlockNotExist
+	}
+	sab, err := node.getCommitSigAndBitmap(blk)
```
Should prefer the commitSig from nextBlock.CommitSig/Bitmap.  Only the last block in the chain rely on the commitSig in DB because it's not saved in the new block yet.

JackyWYX replied:
Yes, this is exactly how currently the logic works. Please take a look into the function `getCommitSigAndBitmap`. 

Also to make the function name more clear, I renamed the function name `getCommitSigAndBitmapFromChildOrDB.`

LeoHChen commented on crypto/hash/rlp.go:
```diff
@@ -1,22 +1,47 @@
 func FromRLP(x interface{}) (h common.Hash) {
-	hw := sha3.NewLegacyKeccak256()
+	hw := kec256Pool.Get().(hash.Hash)
+	defer func() {
```
where is this change coming from?  Is it part of upstream change? why mix it here?

JackyWYX replied:
Ah, this is just to optimize the memory used for RLP hashing in multi-threading. Sorry for mixing it here. Will open another PR for this.

JackyWYX replied:
Reverted and PR here: https://github.com/harmony-one/harmony/pull/3616. Please take a look.

LeoHChen commented on hmy/downloader/shortrange.go:
```diff
@@ -51,21 +49,25 @@ func (d *Downloader) doShortRangeSync() (int, error) {
 			sh.removeStreams(whitelist) // Data provided by remote nodes is corrupted
+		} else {
+			// It is the last block gives a wrong commit sig. Blame the provider of the last block.
+			criminal := stids[len(stids)-1]
```
criminal seems too strong , :-)

JackyWYX replied:
Do you mean the variable name? Renamed to `stToBlame`. Thanks!

LeoHChen commented on internal/chain/engine.go:
```diff
@@ -438,127 +389,113 @@ func applySlashes(
+}
+
+// Support 512 at most validator nodes
+const bitmapKeyBytes = 64
```
at most 512 validators per shard?  Can it be changed later?

JackyWYX replied:
This is simply a sanity field for de-duplication in memory. And sure, it can be changed anytime when needed.

LeoHChen reviewed (commented):
A big change. some quick comments.

rlan35 commented on api/service/legacysync/syncing.go:
```diff
@@ -807,6 +858,12 @@ func (ss *StateSync) UpdateBlockAndStatus(block *types.Block, bc *core.BlockChai
 		Msg("[SYNC] UpdateBlockAndStatus: New Block Added to Blockchain")
+
+	if verifyAllSig && haveCurrentSig {
+		if err := bc.WriteCommitSig(block.NumberU64(), block.GetCurrentCommitSig()); err != nil {
```
no need to explicitly write to DB here. You can just put the commigSigbitmap inside the Block object and the insertChain() will commit the the sig as long as it's present inside the block.

JackyWYX replied:
I feel a little unconfident for writing signature to db that is not verified. Do you think writing unverified commit sig might cause some problem?

JackyWYX replied:
After discussion, it's safe to say this will not be a problem. This field in DB is only used by leader proposing a new block when bootstrap. So if the leader node has a wrong commit sig in DB, it will only effect itself and a view change will be expected later. Removed `bc.WriteCommitSig` in both downloader and DNS sync logic.

rlan35 commented:
overall, looks good.
